src/dtls13.c

```diff
@@ -0,0 +1,1968 @@
+        return BAD_STATE_E;
+    }
+
+    isEncrypted = Dtls13TypeIsEncrypted(handshake_type);
+    rlHeaderLength = Dtls13GetRlHeaderLength(ssl, isEncrypted);
+
+    if (length < rlHeaderLength)
+        return INCOMPLETE_DATA;
+
+    /* DTLSv1.3 do not consider fragmentation for hash transcript. Build the
+       hash now pretending fragmentation will not happen */
+    if (hash_output) {
+        ret = HashRaw(ssl, message + rlHeaderLength, length - rlHeaderLength);
+        if (ret != 0)
+            return ret;
+    }
+
+    messageSize = length - rlHeaderLength - DTLS_HANDSHAKE_HEADER_SZ;
+
+    ssl->dtls13FragmentsBuffer.buffer =
+        (byte *)XMALLOC(messageSize, ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);
+
+    if (ssl->dtls13FragmentsBuffer.buffer == NULL)
+        return MEMORY_E;
+
+    XMEMCPY(ssl->dtls13FragmentsBuffer.buffer,
+        message + rlHeaderLength + DTLS_HANDSHAKE_HEADER_SZ, messageSize);
+
+    ssl->dtls13MessageLength = messageSize;
+    ssl->dtls13FragHandshakeType = handshake_type;
+    ssl->dtls13SendingFragments = 1;
+
+    return Dtls13SendFragmentedInternal(ssl);
+}
+
+/**
+ * dtls13RlAddCiphertextHeader() - add record layer header in the buffer
+ * @ssl: ssl object
+ * @out: output buffer where to put the ehader
+ * @length: length of the record
+ */
+int Dtls13RlAddCiphertextHeader(WOLFSSL *ssl, byte *out, size_t length)
+{
+    Dtls13RecordCiphertextHeader *hdr;
+    word32 seqNumber;
+
+    if (out == NULL)
+        return BAD_FUNC_ARG;
+
+    if (ssl->dtls13EncryptEpoch == NULL)
+        return BAD_STATE_E;
+
+    hdr = (Dtls13RecordCiphertextHeader *)out;
+
+    hdr->unifiedHdrFlags = FIXED_BITS;
+    hdr->unifiedHdrFlags |= (ssl->dtls13EncryptEpoch->epochNumber & EE_MASK);
+
+    /* include 16-bit seq */
+    hdr->unifiedHdrFlags |= S_BIT;
+    /* include 16-bit length */
+    hdr->unifiedHdrFlags |= L_BIT;
+
+    seqNumber = ssl->dtls13EncryptEpoch->nextSeqNumberLo;
+    hdr->sequenceNumber[0] = (seqNumber >> 16) & 0xff;
+    hdr->sequenceNumber[1] = seqNumber & 0xff;
+
+    c16toa(length, hdr->length);
+
+    return 0;
+}
+
+/**
+ * Dtls13HandshakeAddHeader() - add handshake layer header
+ * @ssl: ssl object
+ * @output: output buffer
+ * @msg_type: handshake type
+ * @length: length of the message
+ */
+int Dtls13HandshakeAddheader(
+    WOLFSSL *ssl, byte *output, enum HandShakeType msg_type, size_t length)
+{
+    Dtls13HandshakeHeader *hdr;
+
+    hdr = (Dtls13HandshakeHeader *)output;
+
+    hdr->msg_type = msg_type;
+    c32to24((word32)length, hdr->length);
+    c16toa(ssl->keys.dtls_handshake_number, hdr->messageSeq);
+
+    /* send unfragmented first */
+    c32to24(0, hdr->fragmentOffset);
+    c32to24((word32)length, hdr->fragmentLength);
+
+    return 0;
+}
+
+/**
+ * Dtls13EncryptRecordNumber() - encrypt record number in the header
+ * @ssl: ssl object
+ * @hdr: header
+ *
+ * Further info rfc draft 43 sec 4.2.3
+ */
+int Dtls13EncryptRecordNumber(WOLFSSL *ssl, byte *hdr)
```

Comment:
Somewhere before `Dtls13EncryptRecordNumber` there should be a check that the record length is at least an AES block size. Probably in `BuildTls13Message` but it would be better to pass in the `hdr` length and perform the check here.

Reply:
I added the check here, but somewhere in BuildTls13Message we should implement the possibility to pad the record to satisfy the constrain

Reply:
I agree but that is a future TODO :sunglasses:

Reply:
Hi @julek-wolfssl , this check is added. Do you consider this resolved?